docs/guides/config.md
# Configuration

The Helm chart has two dedicated controllers' configuration sections, pretty well commented in
the [`values.yaml`](https://github.com/alex-karpenko/helm-charts/blob/main/charts/git-events-runner/values.yaml):

* `controllerOptions`: defines values of some controllers' command line options.
* `runtimeConfig`: defines content of the ConfigMap with dynamic controllers' config, or, in other words, with
  overridden defaults.
  It will be reloaded by controller each time it's changed.

Let's explain all of them with some details.

## Controller options

### logLevel

Allowed values: `error`, `warn`, `info`, `debug`, `trace`.

Using this parameter, you can change log level.
If nothing is specified the default level is `info`.

### jsonLogs

Allowed values: `true`, `false`.

Default if `false`.
If `true` then controller outputs logs in JSON format.
JSON format is useful for logging systems like Grafana Loki or ElasticSearch, which expect structured logs.

### otlp

This section describes OpenTelemetry configuration.
If `otlp.enabled` is `true` then `otlp.endpoint` should point to the OpenTelemetry collector endpoint.

### leaderLease

This section declares the parameters of using Lease resource to manage leader elections. There are two parameters:

* `duration`: leader election will be started if the current leaseholder doesn't confirm its lock during this time (in
  seconds).
* `grace`: leaseholder re-confirms (renews) its lock the grace interval (in seconds) before lock expires.

Logically, grace should be less than duration and the grace interval should be enough to re-confirm lock.

### scheduleParallelism, webhooksParallelism

Allowed values: integer from 1 to 255.

These parameters set, for each type of triggers, the maximum number of parallel tasks that the controller runs to watch
for changes in the sources.

> __Important__: This is not a maximum number of simultaneous action jobs.
> This is a way to restrict controller only from running of huge number of simultaneous source verification tasks.
> So there is no way (at least now) to restrict number of action jobs.

### webhooksRateLimits

This section defines Webhooks requests rate limits.
By default, the controller doesn't restrict the rate of the webhook requests,
but rate limiting is a good approach to prevent request flood in some environments,
especially if you're not responsible for configuration of the calling side.

Limits' values are defined as `number-of-requests/period-in-seconds`, for example, "1/10" or "20/60".
But at the same time, the number of requests defines the size of the "burst bucket"
— the number of requests that will be accepted before the rate limiter starts request throttling.
So two equal limits like "1/10" and "10/100" have slightly different meaning:

- 1/10 means that the controller accepts one request and blocks all others during 10 seconds;
- 10/100 means that the controller accepts the first 10 requests and starts blocking all others for 10 seconds, and
  returns one "free seat" into the burst bucket every 10 seconds.

There are three types of limits:

* `global` - limits total webhooks request rate regardless of the trigger of source;
* `trigger` - per trigger limit which restricts rate for each webhook trigger separately, it can't exceed global
  value;
* `source` - per trigger source restriction, it works as previous one but uses each source of each trigger as a rate
  limiting object.

### webhooksTls

This section defines optional TLS config for webhooks server.
If it's omitted (default), webhooks listener works without TLS.

To force webhooks server to use TLS transport, we have to provide a valid server certificate and its private key in PEM
format.
There are two possible ways to do this:

- provide files with certificate and key (or mount Secret volume inside a Pod);
- use existing Secret of the `kubernetes.io/tls` type.

TLS configuration may be provided using the following config parameters:

- `certPath`: path to the file with certificate in PEM format;
- `keyPath`: path to the file with certificate private key in PEM format;
- `secretName`: name of the TLS secret to load certificate and key from;
- `secretNamespace`: namespace where certificate secret is located.

If you specified one of the `certPath` or `keyPath`, the second option becomes mandatory because certificate doesn't
work
without a key and vise versa.

`certPath/keyPath` and `secretName/secretNamespace` are mutually exclusive.
is optional, if it's not specified default namespace will be used: if controller is running inside Kubernetes cluster
(usual way) the default is controllers' namespace, otherwise namespace `default` will be used.

### secretsCacheTime

This parameter specifies the maximum number of seconds to hold values in the cache of secrets.

Controller resolves content of the secrets each time it needs secret value (for example, WebhookTrigger can use auth
token stored in a secret).
To eliminate Kubernetes API overloading by huge number of requests for secret values, the controller has some kind of
shared cache for secrets.
It holds only vales used by controller.

### sourceCloneFolder

This section defines which folder (`mountPath`) will be used inside the controller to clone content of the sources
during periodic
verification for changes.

At the same time, parameter `volumeName` defines name of the Pods' volume, which should be used for that folder.
Helm chart defines controllers' Pod with this volume of type `emptyDir`.

### metricsPrefix

The parameter specifies prefix ("namespace") to use for Prometheus metrics.
The default value is `git_events_runner`.

## Runtime config

### trigger

Defines triggers defaults:

* `webhook.defaultAuthHeader`: default header name with authentication token in a WebhookTrigger request. May be changed
  in the trigger config. Default is `x-trigger-auth`.

### action

This section defines lots of defaults of action jobs.

| Parameter name                    | Default value                                                                  | Description                                                                                                                                |
| --------------------------------- | ------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------ |
| ttlSecondsAfterFinished           | 7200                                                                           | Default time to leave of Job after finishing. After this time Jobs will be removed from the cluster with its Pod. Useful fo debug purpose. |
| activeDeadlineSeconds             | 3600                                                                           | Default time limit to run Job. After this time incomplete Job will be terminated.                                                          |
| maxRunningJobs                    | 16                                                                             | Maximum number of simultaneously running Jobs, per controller replica. Jobs that can't be running will be queued and waiting.              |
| jobWaitingTimeoutSeconds          | 300                                                                            | Maximum time (in seconds) Jobs can wait in queue before start because `maxRunningJobs` is exceeded.                                        |
| defaultServiceAccount             | `{{ fullname }}-action-job`                                                    | Default service account name for action jobs. Actual default depends on release name and `fullnameOverride` global parameter.              |
| workdir.mountPath                 | /action_workdir                                                                | Default folder to clone source content to. It's used for both cloner and worker container.                                                 |
| workdir.volumeName                | action-workdir                                                                 | Volume name of workdir `emptyDir` volume.                                                                                                  |
| containers.cloner.name            | action-cloner                                                                  | Name of the source cloner initContainer in the action Job.                                                                                 |
| containers.cloner.image           | ghcr.io/alex-karpenko/git-events-runner/gitrepo-cloner:{{ .Chart.AppVersion }} | Default image to use for source cloner container.                                                                                          |
| containers.worker.name            | action-worker                                                                  | Name of the action worker container in the action Job.                                                                                     |
| containers.worker.image           | ghcr.io/alex-karpenko/git-events-runner/action-worker:{{ .Chart.AppVersion }}  | Default image to use for the action worker container.                                                                                      |
| containers.worker.variablesPrefix | ACTION_JOB_                                                                    | Prefix for environment variables of the action worker container, that controller creates to pass trigger and sources parameters.           |

## Command line parameters

Controller application has three subcommands:

* `crds`: to print CRD manifests to the stdout.
  It's useful to install CRD declarations to your cluster directly.
  It has no additional options.
* `config`: to print a default dynamic config to the stdout.
  It's useful to get base config and modify parts you want to customize.
  Has only one optional parameter `--helm-template` to dump some templates instead of default values:
  this option is useful in CI/CD pipelines to automate Helm chart linting and templating.
* `run`: to run controller, all command line parameters are optional and have defaults.
  All of them are described as a static configuration options above in this section.

```bash
docker run --rm ghcr.io/alex-karpenko/git-events-runner/git-events-runner:latest --help

Kubernetes operator to run Jobs on events from Git

Usage: git-events-runner <COMMAND>

Commands:
  crds    Print CRD definitions to stdout
  config  Print default dynamic config YAML to stdout
  run     Run K8s controller
  help    Print this message or the help of the given subcommand(s)

Options:
  -h, --help     Print help
  -V, --version  Print version
```

```bash
docker run --rm ghcr.io/alex-karpenko/git-events-runner/git-events-runner:latest run --help

Run K8s controller

Usage: git-events-runner run [OPTIONS]

Options:
  -w, --webhooks-port <WEBHOOKS_PORT>
          Port to listen on for webhooks [default: 8080]
  -u, --utility-port <UTILITY_PORT>
          Port to listen on for utilities web [default: 3000]
      --webhooks-parallelism <WEBHOOKS_PARALLELISM>
          Maximum number of webhook triggers running in parallel [default: 16]
      --schedule-parallelism <SCHEDULE_PARALLELISM>
          Maximum number of schedule triggers running in parallel [default: 16]
      --secrets-cache-time <SECRETS_CACHE_TIME>
          Seconds to cache secrets for [default: 60]
      --source-clone-folder <SOURCE_CLONE_FOLDER>
          Path (within container) to clone repo to [default: /tmp/git-events-runner]
      --config-map-name <CONFIG_MAP_NAME>
          Name of the ConfigMap with dynamic controller config [default: git-events-runner-config]
      --leader-lease-name <LEADER_LEASE_NAME>
          Name of the Lease for leader locking [default: git-events-runner-leader-lock]
      --leader-lease-duration <LEADER_LEASE_DURATION>
          Leader lease duration, seconds [default: 30]
      --leader-lease-grace <LEADER_LEASE_GRACE>
          Leader lease grace interval, seconds [default: 5]
      --metrics-prefix <METRICS_PREFIX>
          Name of the ConfigMap with dynamic controller config [default: git_events_runner]
      --hooks-rrl-global <HOOKS_RRL_GLOBAL>
          Global webhooks requests rate limit (burst limit/seconds)
      --hooks-rrl-trigger <HOOKS_RRL_TRIGGER>
          Requests rate limit (burst limit/seconds) per webhook trigger
      --hooks-rrl-source <HOOKS_RRL_SOURCE>
          Requests rate limit (burst limit/seconds) per webhook source
      --tls-cert-path <TLS_CERT>
          Path to TLS certificate file
      --tls-key-path <TLS_KEY>
          Path to TLS key file
      --tls-secret-name <TLS_SECRET_NAME>
          Secret name with TLS certificate and key
      --tls-secret-namespace <TLS_SECRET_NAMESPACE>
          Namespace of the  TLS secret
  -j, --json-logs
          Output logs in JSON format
  -h, --help
          Print help
```

## gitrepo-cloner options

Usually you don't use this image (or app) directly, but the controller runs Jobs with `gitrepo-cloner` init image.
It doesn't have configuration files but command line options only.

Just for the sake of curiosity or for debugging purpose:

```bash
docker run --rm ghcr.io/alex-karpenko/git-events-runner/gitrepo-cloner:latest --help

Git repo cloner, supplementary tool for git-events-runner operator

Usage: gitrepo-cloner [OPTIONS] --kind <SOURCE_KIND> --source <SOURCE_NAME> --destination <DESTINATION> <--branch <BRANCH>|--tag <TAG>|--commit <COMMIT>>

Options:
  -k, --kind <SOURCE_KIND>         Source kind
  -s, --source <SOURCE_NAME>       Source name
  -b, --branch <BRANCH>            Branch name
  -t, --tag <TAG>                  Tag name
  -c, --commit <COMMIT>            Commit hash
  -d, --destination <DESTINATION>  Destination folder
  -p, --preserve-git-folder        Don't remove .git folder after clone
      --debug                      Set log level to debug
  -h, --help                       Print help
  -V, --version                    Print version
```

> Note: `--branch`, `--tag` and `--commit` options are mutually exclusive, but one of them is mandatory;


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.3.1] - 2024-10-01
### Details
#### Changed
- Logs in JSON format by @alex-karpenko

## [0.3.0] - 2024-09-28
### Details
#### Added
- Implement web-hooks requests rate limiting by @alex-karpenko in [#35](https://github.com/alex-karpenko/git-events-runner/pull/35)
- Implement TLS listener for webhooks by @alex-karpenko in [#36](https://github.com/alex-karpenko/git-events-runner/pull/36)

## [0.2.4] - 2024-09-21
### Details
#### Changed
- Update dependencies by @alex-karpenko in [#30](https://github.com/alex-karpenko/git-events-runner/pull/30)
- Improve documentation of the code by @alex-karpenko in [#32](https://github.com/alex-karpenko/git-events-runner/pull/32)
- Update dependencies by @alex-karpenko in [#33](https://github.com/alex-karpenko/git-events-runner/pull/33)
- Update kube to v0.95, k8s-openapi to v0.23, kube-lease-manager to v0.4 by @alex-karpenko in [#34](https://github.com/alex-karpenko/git-events-runner/pull/34)

#### Fixed
- Fix flapping test reconcile_schedule_trigger_should_set_idle_status by @alex-karpenko in [#31](https://github.com/alex-karpenko/git-events-runner/pull/31)

## [0.2.3] - 2024-07-19
### Details
#### Fixed
- Update dependencies to fix security findings by @alex-karpenko in [#29](https://github.com/alex-karpenko/git-events-runner/pull/29)

## [0.2.2] - 2024-07-08
### Details
#### Fixed
- Fix large default leader lock grace interval by @alex-karpenko in [#25](https://github.com/alex-karpenko/git-events-runner/pull/25)
- Update to use kube-lease-manager instead of kubert crate by @alex-karpenko in [#28](https://github.com/alex-karpenko/git-events-runner/pull/28)

## [0.2.1] - 2024-06-24
### Details
#### Added
- Implement runtime metrics and imorove traces by @alex-karpenko in [#23](https://github.com/alex-karpenko/git-events-runner/pull/23)

#### Changed
- Improve unit and integration testing by @alex-karpenko in [#21](https://github.com/alex-karpenko/git-events-runner/pull/21)

#### Fixed
- Fix security findings by @alex-karpenko in [#22](https://github.com/alex-karpenko/git-events-runner/pull/22)
- Update dependencies by @alex-karpenko in [#24](https://github.com/alex-karpenko/git-events-runner/pull/24)

## [0.2.0] - 2024-06-15
### Details
#### Breaking changes
- Implement glob file pattern matching instead of the single file name specification in the triggers by @alex-karpenko in [#14](https://github.com/alex-karpenko/git-events-runner/pull/14)

#### Added
- Implement actionJob config with additional labels and annotations by @alex-karpenko in [#10](https://github.com/alex-karpenko/git-events-runner/pull/10)
- Implement actinJob config with affinity, nodeSelector and tolerations rules by @alex-karpenko in [#11](https://github.com/alex-karpenko/git-events-runner/pull/11)
- Implement job execution time limit (active_deadline_seconds) by @alex-karpenko in [#12](https://github.com/alex-karpenko/git-events-runner/pull/12)
- Implement limitation of the running jobs number by @alex-karpenko in [#13](https://github.com/alex-karpenko/git-events-runner/pull/13)
- Add jobs requeueing when config changed by @alex-karpenko in [#17](https://github.com/alex-karpenko/git-events-runner/pull/17)

#### Changed
- Make documentation versioned by @alex-karpenko in [#15](https://github.com/alex-karpenko/git-events-runner/pull/15)
- Improve logging, make it more formal and short, with relevant info only by @alex-karpenko in [#16](https://github.com/alex-karpenko/git-events-runner/pull/16)

#### Fixed
- Update documentation to reflect changes in Helm chart by @alex-karpenko in [#9](https://github.com/alex-karpenko/git-events-runner/pull/9)
- Documentation deployment workflow by @alex-karpenko
- Fix grammar, syntax and lint warnings by @alex-karpenko in [#19](https://github.com/alex-karpenko/git-events-runner/pull/19)

## [0.1.1] - 2024-06-01
### Details
#### Added
- Add subcommand to dump default dynamic config by @alex-karpenko in [#8](https://github.com/alex-karpenko/git-events-runner/pull/8)

#### Changed
- Bump requests from 2.31.0 to 2.32.2 by @dependabot[bot] in [#4](https://github.com/alex-karpenko/git-events-runner/pull/4)
- Push images to docker.io registry by @alex-karpenko in [#6](https://github.com/alex-karpenko/git-events-runner/pull/6)

#### Fixed
- Fix links in the README by @alex-karpenko
- Change action-worker base image to Ubuntu 24.04 by @alex-karpenko in [#7](https://github.com/alex-karpenko/git-events-runner/pull/7)

## New Contributors
* @dependabot[bot] made their first contribution in [#4](https://github.com/alex-karpenko/git-events-runner/pull/4)

## [0.1.0] - 2024-05-24
### Details
#### Changed
- First beta implementation by @alex-karpenko in [#2](https://github.com/alex-karpenko/git-events-runner/pull/2)

#### Fixed
- Fix error in release workflow by @alex-karpenko in [#5](https://github.com/alex-karpenko/git-events-runner/pull/5)
- Fix release workflow permissions by @alex-karpenko

#### Removed
- Remove changelog updating from the release workflow by @alex-karpenko

[0.3.1]: https://github.com/alex-karpenko/git-events-runner/compare/v0.3.0..v0.3.1
[0.3.0]: https://github.com/alex-karpenko/git-events-runner/compare/v0.2.4..v0.3.0
[0.2.4]: https://github.com/alex-karpenko/git-events-runner/compare/v0.2.3..v0.2.4
[0.2.3]: https://github.com/alex-karpenko/git-events-runner/compare/v0.2.2..v0.2.3
[0.2.2]: https://github.com/alex-karpenko/git-events-runner/compare/v0.2.1..v0.2.2
[0.2.1]: https://github.com/alex-karpenko/git-events-runner/compare/v0.2.0..v0.2.1
[0.2.0]: https://github.com/alex-karpenko/git-events-runner/compare/v0.1.1..v0.2.0
[0.1.1]: https://github.com/alex-karpenko/git-events-runner/compare/v0.1.0..v0.1.1

<!-- generated by git-cliff -->


src/cli.rs
//! Cli parameters for git-events-runner binary
use anyhow::Context;
use axum_server::tls_rustls::RustlsConfig;
use clap::Parser;
use k8s_openapi::api::core::v1::Secret;
use kube::{Api, Client};
use std::sync::OnceLock;
use tracing::debug;

use crate::{web::RequestsRateLimitParams, Error};

const DEFAULT_SOURCE_CLONE_FOLDER: &str = "/tmp/git-events-runner";
const DEFAULT_CONFIG_MAP_NAME: &str = "git-events-runner-config";
const DEFAULT_LEADER_LOCK_LEASE_NAME: &str = "git-events-runner-leader-lock";
const DEFAULT_METRICS_PREFIX: &str = "git_events_runner";

static CLI_CONFIG: OnceLock<CliConfig> = OnceLock::new();

/// Root CLI commands
#[derive(Parser, Debug)]
#[command(author, version, about)]
#[allow(clippy::large_enum_variant)]
pub enum Cli {
    /// Print CRD definitions to stdout
    Crds,
    /// Print default dynamic config YAML to stdout
    Config(CliConfigDumpOptions),
    /// Run K8s controller
    Run(CliConfig),
}

/// Parameters for the `run` subcommand
#[derive(Parser, Debug, Clone)]
pub struct CliConfig {
    /// Port to listen on for webhooks
    #[arg(long, short, value_parser=clap::value_parser!(u16).range(1..), default_value = "8080")]
    pub webhooks_port: u16,

    /// Port to listen on for utilities web
    #[arg(long, short, value_parser=clap::value_parser!(u16).range(1..), default_value = "3000")]
    pub utility_port: u16,

    /// Maximum number of webhook triggers running in parallel
    #[arg(long, value_parser=clap::value_parser!(u16).range(1..256), default_value = "16")]
    pub webhooks_parallelism: u16,

    /// Maximum number of schedule triggers running in parallel
    #[arg(long, value_parser=clap::value_parser!(u16).range(1..256), default_value = "16")]
    pub schedule_parallelism: u16,

    /// Seconds to cache secrets for
    #[arg(long, value_parser=clap::value_parser!(u64).range(1..), default_value = "60")]
    pub secrets_cache_time: u64,

    /// Path (within container) to clone repo to
    #[arg(long, default_value = DEFAULT_SOURCE_CLONE_FOLDER)]
    pub source_clone_folder: String,

    /// Name of the ConfigMap with dynamic controller config
    #[arg(long, default_value = DEFAULT_CONFIG_MAP_NAME)]
    pub config_map_name: String,

    /// Name of the Lease for leader locking
    #[arg(long, default_value = DEFAULT_LEADER_LOCK_LEASE_NAME)]
    pub leader_lease_name: String,

    /// Leader lease duration, seconds
    #[arg(long, value_parser=clap::value_parser!(u64).range(1..301), default_value = "30")]
    pub leader_lease_duration: u64,

    /// Leader lease grace interval, seconds
    #[arg(long, value_parser=clap::value_parser!(u64).range(1..301), default_value = "5")]
    pub leader_lease_grace: u64,

    /// Name of the ConfigMap with dynamic controller config
    #[arg(long, default_value = DEFAULT_METRICS_PREFIX)]
    pub metrics_prefix: String,

    /// Global webhooks requests rate limit (burst limit/seconds)
    #[arg(long, value_parser=Cli::rrl_parser)]
    pub hooks_rrl_global: Option<RequestsRateLimitParams>,

    /// Requests rate limit (burst limit/seconds) per webhook trigger
    #[arg(long, value_parser=Cli::rrl_parser)]
    pub hooks_rrl_trigger: Option<RequestsRateLimitParams>,

    /// Requests rate limit (burst limit/seconds) per webhook source
    #[arg(long, value_parser=Cli::rrl_parser)]
    pub hooks_rrl_source: Option<RequestsRateLimitParams>,

    /// Path to the TLS certificate file
    #[arg(long, requires = "tls_key_path", conflicts_with_all=["tls_secret_name", "tls_secret_namespace"])]
    pub tls_cert_path: Option<String>,

    /// Path to the TLS key file
    #[arg(long, requires = "tls_cert_path", conflicts_with_all=["tls_secret_name", "tls_secret_namespace"])]
    pub tls_key_path: Option<String>,

    /// Name of the Secret with TLS certificate and key
    #[arg(long, conflicts_with_all = ["tls_cert_path", "tls_key_path"])]
    pub tls_secret_name: Option<String>,

    /// Namespace of the TLS secret
    #[arg(long, requires = "tls_secret_name", conflicts_with_all = ["tls_cert_path", "tls_key_path"])]
    pub tls_secret_namespace: Option<String>,

    /// Output logs in JSON format
    #[arg(long, short = 'j')]
    pub json_logs: bool,
}

/// Parameters for the `config` subcommand
#[derive(Parser, Debug, Clone)]
pub struct CliConfigDumpOptions {
    /// Include some templates for Helm chart
    #[arg(short = 't', long)]
    pub helm_template: bool,
}

impl Cli {
    /// Constructs CLI config
    #[allow(clippy::new_without_default)]
    pub fn new() -> Cli {
        let cli: Cli = Parser::parse();
        debug!(config = ?cli, "creating cli config");

        // If we run controller - set shared CLiConfig instance
        if let Cli::Run(config) = cli {
            CLI_CONFIG.set(config.clone()).unwrap();
            Cli::Run(config)
        } else {
            cli
        }
    }

    /// Parse RRL params using original try_from(&str)
    fn rrl_parser(s: &str) -> Result<RequestsRateLimitParams, String> {
        RequestsRateLimitParams::try_from(s).map_err(|e| e.to_string())
    }
}

impl CliConfig {
    /// Returns really parsed CLI parameters or test mock with defaults
    pub fn get() -> &'static Self {
        #[cfg(not(test))]
        {
            CLI_CONFIG.get().unwrap()
        }

        #[cfg(test)]
        {
            CLI_CONFIG.get_or_init(|| CliConfig::parse_from::<_, &String>(&[]))
        }
    }

    /// Returns ready to use RustlsConfig instance if TLS config options were provided,
    /// or None if no config
    pub async fn build_tls_config(&self, client: Client) -> anyhow::Result<Option<RustlsConfig>> {
        match (
            self.tls_cert_path.as_ref(),
            self.tls_key_path.as_ref(),
            self.tls_secret_name.as_ref(),
        ) {
            (Some(cert), Some(key), _) => {
                let config = RustlsConfig::from_pem_file(cert, key)
                    .await
                    .context(format!("Unable to apply TLS config with cert={cert} and key={key}"))?;
                Ok(Some(config))
            }
            (_, _, Some(secret_name)) => {
                let secret_api: Api<Secret> = if let Some(namespace) = self.tls_secret_namespace.as_ref() {
                    Api::namespaced(client, namespace)
                } else {
                    Api::default_namespaced(client)
                };
                let secret = secret_api
                    .get(secret_name)
                    .await
                    .context(format!("Unable to get TLS secret {secret_name}"))?;
                let secret_data = secret.data.ok_or_else(|| {
                    Error::SecretDecodingError(format!("no `data` part in the secret `{}`", secret_name))
                })?;
                let cert = secret_data
                    .get("tls.crt")
                    .ok_or_else(|| {
                        Error::SecretDecodingError(format!("no `tls.crt` key in the secret `{}`", secret_name))
                    })?
                    .to_owned();
                let cert = String::from_utf8(cert.0)
                    .context("TLS cert is not a valid UTF-8 string")?
                    .into_bytes();

                let key = secret_data
                    .get("tls.key")
                    .ok_or_else(|| {
                        Error::SecretDecodingError(format!("no `tls.key` key in the secret `{}`", secret_name))
                    })?
                    .to_owned();
                let key = String::from_utf8(key.0)
                    .context("TLS key is not a valid UTF-8 string")?
                    .into_bytes();

                let config = RustlsConfig::from_pem(cert, key).await?;
                Ok(Some(config))
            }
            (_, _, _) => Ok(None),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tests;

    use insta::assert_debug_snapshot;
    use k8s_openapi::{api::core::v1::Namespace, ByteString};
    use kube::{
        api::{DeleteParams, ObjectMeta, PostParams},
        Resource as _,
    };
    use std::{collections::BTreeMap, env};
    use tokio::{fs::File, io::AsyncReadExt, sync::OnceCell};

    const NAMESPACE: &str = "secret-tls-test";
    const TEST_CERTIFICATES_SECRET_NAME: &str = "test-tls-certificates";
    const SERVER_CERT_BUNDLE: &str = "/test-server.pem";
    const SERVER_END_CERT: &str = "/end.crt";
    const SERVER_PRIVATE_KEY: &str = "/test-server.key";

    static INITIALIZED: OnceCell<()> = OnceCell::const_new();

    async fn init(client: Client) {
        INITIALIZED
            .get_or_init(|| async {
                tests::init_crypto_provider().await;
                create_namespace(client).await;
            })
            .await;
    }

    /// Unattended namespace creation
    async fn create_namespace(client: Client) {
        let api = Api::<Namespace>::all(client);
        let pp = PostParams::default();

        let mut data = Namespace::default();
        data.meta_mut().name = Some(String::from(NAMESPACE));
        api.create(&pp, &data).await.unwrap_or_default();
    }

    #[test]
    fn default_cli_consistency() {
        let cli = CliConfig::parse_from::<_, &str>([]);
        assert_debug_snapshot!(cli);
    }

    #[tokio::test]
    #[ignore = "needs docker"]
    async fn tls_config_nothing() {
        tests::init_crypto_provider().await;
        let client = tests::get_test_kube_client().await.unwrap();

        let cli = CliConfig::parse_from::<_, &str>([]);
        let config = cli.build_tls_config(client).await.unwrap();
        assert!(config.is_none());
    }

    #[tokio::test]
    #[ignore = "needs docker"]
    async fn tls_config_cert_key() {
        let client = tests::get_test_kube_client().await.unwrap();
        init(client.clone()).await;

        let out_dir = env::var("OUT_DIR").unwrap();
        let cert_bundle_path = format!("{out_dir}/tls/{SERVER_CERT_BUNDLE}");
        let cert_path = format!("{out_dir}/tls/{SERVER_END_CERT}");
        let key_path = format!("{out_dir}/tls/{SERVER_PRIVATE_KEY}");

        let cli = CliConfig::parse_from::<_, &str>([
            "run",
            "--tls-cert-path",
            cert_bundle_path.as_str(),
            "--tls-key-path",
            key_path.as_str(),
        ]);
        let config = cli.build_tls_config(client.clone()).await.unwrap();
        assert!(config.is_some());

        let cli = CliConfig::parse_from::<_, &str>([
            "run",
            "--tls-cert-path",
            cert_path.as_str(),
            "--tls-key-path",
            key_path.as_str(),
        ]);
        let config = cli.build_tls_config(client).await.unwrap();
        assert!(config.is_some());
    }

    #[tokio::test]
    #[ignore = "needs docker"]
    async fn tls_config_secret() {
        let client = tests::get_test_kube_client().await.unwrap();
        init(client.clone()).await;

        let out_dir = env::var("OUT_DIR").unwrap();
        let cert_bundle_path = format!("{out_dir}/tls/{SERVER_CERT_BUNDLE}");
        let cert_path = format!("{out_dir}/tls/{SERVER_END_CERT}");
        let key_path = format!("{out_dir}/tls/{SERVER_PRIVATE_KEY}");

        let cli = CliConfig::parse_from::<_, &str>([
            "run",
            "--tls-secret-name",
            TEST_CERTIFICATES_SECRET_NAME,
            "--tls-secret-namespace",
            NAMESPACE,
        ]);
        let secret_api: Api<Secret> = Api::namespaced(client.clone(), NAMESPACE);

        // Delete possible leftovers from the previous run
        let _ = secret_api
            .delete(TEST_CERTIFICATES_SECRET_NAME, &DeleteParams::default())
            .await;

        // Test with bundle
        let mut cert = vec![];
        let mut key = vec![];
        File::open(cert_bundle_path)
            .await
            .unwrap()
            .read_to_end(&mut cert)
            .await
            .unwrap();
        File::open(key_path).await.unwrap().read_to_end(&mut key).await.unwrap();

        let cert: ByteString = ByteString(cert);
        let key: ByteString = ByteString(key);

        let secret = Secret {
            type_: Some("kubernetes.io/tls".to_string()),
            metadata: ObjectMeta {
                name: Some(TEST_CERTIFICATES_SECRET_NAME.to_string()),
                namespace: Some(NAMESPACE.to_string()),
                ..ObjectMeta::default()
            },
            data: Some(BTreeMap::from([
                ("tls.crt".to_string(), cert),
                ("tls.key".to_string(), key.clone()),
            ])),
            ..Secret::default()
        };

        secret_api.create(&PostParams::default(), &secret).await.unwrap();

        let config = cli.build_tls_config(client.clone()).await.unwrap();
        assert!(config.is_some());

        secret_api
            .delete(TEST_CERTIFICATES_SECRET_NAME, &DeleteParams::default())
            .await
            .unwrap();

        // Test end certificate
        let mut cert = vec![];
        File::open(cert_path)
            .await
            .unwrap()
            .read_to_end(&mut cert)
            .await
            .unwrap();

        let cert: ByteString = ByteString(cert);

        let secret = Secret {
            type_: Some("kubernetes.io/tls".to_string()),
            metadata: ObjectMeta {
                name: Some(TEST_CERTIFICATES_SECRET_NAME.to_string()),
                namespace: Some(NAMESPACE.to_string()),
                ..ObjectMeta::default()
            },
            data: Some(BTreeMap::from([
                ("tls.crt".to_string(), cert),
                ("tls.key".to_string(), key),
            ])),
            ..Secret::default()
        };

        secret_api.create(&PostParams::default(), &secret).await.unwrap();

        let config = cli.build_tls_config(client).await.unwrap();
        assert!(config.is_some());

        secret_api
            .delete(TEST_CERTIFICATES_SECRET_NAME, &DeleteParams::default())
            .await
            .unwrap();
    }
}
